### PR TITLE
Implement Semgrep provider

### PR DESCRIPTION
**Pull request by tinvaan**
Supports @-mentions.

https://github.com/sourcegraph/openctx/assets/5815030/b3d88e69-bacf-47dd-a7b6-65c6ac814623

Click [here](https://youtu.be/Cm_9QGrbMdA) for high-res screen recording.

### TODO
- Fix annotations
- Add more tests
- Publish npm package for provider

### CODE REVIEW

File: provider/semgrep/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+This is a sample context provider for [OpenCtx](https://openctx.org) that fetches the contents of a Semgrep finding from a given URL.
+
+## Usage
+
+Add the following to your settings in any OpenCtx client:
+
+```json
+"openctx.providers": {
+    // ...other providers...
+    "https://openctx.org/npm/@openctx/provider-semgrep": {
```
Comment (tinvaan) [inline provider/semgrep/README.md:12]: @sqs can you help me publish this package under https://openctx.org/npm/@openctx ?

File: provider/semgrep/index.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        finding.rule_message ?? '',
+        finding.rule.name ?? '',
+        finding.rule.message ?? '',
+        finding.triage_comment ?? '',
+        finding.triage_reason ?? '',
+        finding.assistant?.autofix?.fix_code ?? '',
+        finding.assistant?.autofix?.explanation ?? '',
+        finding.assistant?.autotriage?.reason ?? '',
+    ].join('\n')
+}
```
Comment (tinvaan) [inline provider/semgrep/index.ts:48]: not exactly sure if this is the right way to pass a prompt to `ai` in the `items()` method. are there any guides? looking for feedback.

File: provider/semgrep/package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@openctx/provider-semgrep",
+  "version": "0.0.12",
```
Comment (eseliger) [inline provider/semgrep/package.json:3]: if this was never published, should this be 0.0.1 instead?

File: provider/semgrep/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        "deployment": "Semgrep deployment",
+        "repo": "Semgrep repository or project filter",
+        "token": "Your <SEMGREP_APP_TOKEN>"",
```
Comment (eseliger) [inline provider/semgrep/README.md:15]: can you add a quick paragraph on how to acquire such a token?

File: provider/semgrep/client.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        this.repo = settings.repo || ''
+        this.deployment = settings.deployment
+        this.url = 'https://semgrep.dev/api/v1'
+    }
+
+    private abort(): any {
+        if (this.signal === null) {
+            const abc = new AbortController()
+            setTimeout(() => abc.abort(), 5000)
+            this.signal = abc.signal
```
Comment (eseliger) [inline provider/semgrep/client.ts:23]: Would reusing the abortcontroller here mean that it would be cancelled for followup requests? (not a huge expert with abortcontrollers)

File: provider/semgrep/index.ts (hunk 2 above)
Comment (eseliger) [reply provider/semgrep/index.ts:48]: I see that for many providers we used XML, but we don't seem to be super consistent here.
Might be worth a shot? Look for 'fast-xml-parser'

**Review by eseliger — COMMENTED** [review]
nothing stands out, also pinged noodle for another review (and so they can release the package and merge the PR)

File: provider/semgrep/package.json (hunk 3 above)
Comment (tinvaan) [reply provider/semgrep/package.json:3]: lol yes, I copied it from the `hello-world` provider directory. amending.

File: provider/semgrep/client.ts (hunk 5 above)
Comment (tinvaan) [reply provider/semgrep/client.ts:23]: honestly, I'm not sure either. will do some reading up re this.

Comment (tinvaan) [reply provider/semgrep/client.ts]: looks like reusing abortcontrollers is a bad idea -- reverting.

File: provider/semgrep/index.ts (hunk 2 above)
Comment (tinvaan) [reply provider/semgrep/index.ts]: ok, I've also gone with the XML structure purely because it looks more intuitive than just exposing a bunch of raw text, but perhaps we should add some examples somewhere in the docs for `AssistantInfo.content?` property might look like?

/cc @sqs

**Comment by tinvaan** [comment]
@eseliger incorporated your feedback, thanks!

File: provider/semgrep/index.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    },
+
+    async mentions(params: MentionsParams, settings: Settings): Promise<MentionsResult> {
+        if (params.query) {
+            const q: { deployment: string; finding: number } | null = parseUri(params.query)
+            if (q) {
+                const client: API = new API({
+                    repo: settings.repo,
+                    token: settings.token,
+                    deployment: q.deployment,
```
Comment (keegancsmith) [inline provider/semgrep/index.ts:89]: here you use the deployment that was parsed out of the params.query (rather than from settings). But then in items you use settings.deployment for the URL. I assume that is incorrect in items? Instead you should be parsing out the deployment from the mention.uri.

Is the only point of settings.deployment for annotations? If so I would consider making it optional, or if removing annotations support remove deployments from the config.

File: provider/semgrep/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "../../.config/tsconfig.base.json",
+  "esModuleInterop": true,
+  "compilerOptions": {
+    "rootDir": ".",
+    "outDir": "dist",
+    "lib": ["ESNext", "DOM"],
+  },
+  "include": ["*.ts"],
+  "exclude": ["dist", "*.test.*", "vitest.config.ts"],
```
Comment (keegancsmith) [inline provider/semgrep/tsconfig.json:10]: need to exclude mock dir?

**Review by keegancsmith — CHANGES_REQUESTED** [review]
nice work. Just have some inline comments around annotations and if we should be shipping them. Additionally a bug around deployments in urls.

File: provider/semgrep/index.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                          title: f.rule_name,
+                          data: { finding: f },
+                          uri: urlfor(q.deployment, f.repository.name, f.id),
+                      }))
+            }
+        }
+        return []
+    },
+
+    async annotations(params: AnnotationsParams, settings: Settings): Promise<AnnotationsResult> {
```
[the remark this answers is not in the document]
Comment (tinvaan) [reply provider/semgrep/index.ts:104]: Yeah, annotations are more of a PoC at this point, agree it's probably best to ship it in a later release. I'll take it down & raise a ticket in the issue tracker.

[the remark this answers is not in the document]
Comment (tinvaan) [reply provider/semgrep/index.ts:104]: Tracking here: https://github.com/sourcegraph/openctx/issues/125

File: provider/semgrep/index.ts (hunk 6 above)
Comment (tinvaan) [reply provider/semgrep/index.ts:89]: Thanks, that's a good spot. Yeah, `settings.deployment` is more useful for annotations. Nevertheless, I'll make it optional and use it as a fallback in case `mentions.uri` fails to populate for some reason.

Comment (tinvaan) [reply provider/semgrep/index.ts:88]: Addressed in https://github.com/sourcegraph/openctx/pull/105/commits/d2cd580b69a3d18a0a4ae7b156f79269b800377f & https://github.com/sourcegraph/openctx/pull/105/commits/70961707541974ebc922e10e70ebbc55395f4d68

File: provider/semgrep/index.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    Provider,
+} from '@openctx/provider'
+import { XMLBuilder } from 'fast-xml-parser'
+
+import { type Finding, type Findings, urlfor } from './api.js'
+import API from './client.js'
+
+export type Settings = {
+    token: string // Semgrep app token
+    repo?: string // Semgrep scan repository
```
[the remark this answers is not in the document]
Comment (tinvaan) [reply provider/semgrep/index.ts:17]: that's correct, `repo` acts as a filter. I will add a line in the README.

File: provider/semgrep/index.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    async mentions(params: MentionsParams, settings: Settings): Promise<MentionsResult> {
+        if (params.query) {
+            const q: { deployment: string; finding: number } | null = parseUri(params.query)
+            if (q) {
+                const client: API = new API({
+                    repo: settings.repo,
+                    token: settings.token,
+                    deployment: q.deployment,
+                })
+                const findings: Findings = await client.findings(q.finding)
```
[the remark this answers is not in the document]
Comment (tinvaan) [reply provider/semgrep/index.ts:90]: Yeah, I just checked the API docs again & I think I can readily use a q filter like `?repos=${this.deployment}/${this.repo}` available on [`GET /deployments/<deployment slug>/findings`](https://semgrep.dev/api/v1/docs/#tag/Finding/operation/semgrep_app.core_exp.findings.handlers.issue.openapi_list_recent_issues)

I think I had been using the q filters on the API endpoint wrong & hence had to do the `.filter()` on the client side. Let me try again.

[the remark this answers is not in the document]
Comment (tinvaan) [reply provider/semgrep/index.ts:90]: Not sure what I'm missing, the query filter returns empty.
```shell
➞  curl -X GET https://semgrep.dev/api/v1/deployments -H "Authorization: Bearer <token>" | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   142  100   142    0     0    177      0 --:--:-- --:--:-- --:--:--   177
{
  "deployments": [
    {
      "id": 26671,
      "name": "tinvaan",
      "slug": "tinvaan",
      "findings": {
        "url": "https://semgrep.dev/api/v1/deployments/tinvaan/findings"
      }
    }
  ]
}


➞  curl -X GET https://semgrep.dev/api/v1/deployments/tinvaan/findings\?repos\="tinvaan/opencodegraph" -H "Authorization: Bearer <token>" | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    16  100    16    0     0     19      0 --:--:-- --:--:-- --:--:--    19
{
  "findings": []
}
```

I think I'll leave the logic as is and tackle it in a separate PR.

**Comment by tinvaan** [comment]
hey @keegancsmith can you help merge & publish, if all ok? Thanks!

**Comment by keegancsmith** [comment]
https://www.npmjs.com/package/@openctx/provider-semgrep :)

**Comment by tinvaan** [comment]
Thank you! <3
